src/accumulation.jl

```diff
@@ -6,25 +6,35 @@ This avoids allocations when `x` can be mutated in this way.
 """
 add!!(x, y) = x + y
 
-"""
-    add!!(x, t::InplacableThunk)
-
-The specialization of `add!!` for [`InplaceableThunk`](@ref) promises to only call
-`t.add!` on `x` if `x` is suitably mutable; otherwise it will be out of place.
-"""
 function add!!(x, t::InplaceableThunk)
     return if is_inplaceable_destination(x)
-        if !debug_mode()
-            t.add!(x)
-        else
+        if debug_mode()
             debug_add!(x, t)
+        else
+            t.add!(x)
         end
     else
-        x + t
+        # Can't mutate x, but this may still be more efficient:
+        add!!(unthunk(t), x)
```

Comment:
What do you think of
```suggestion
        # Fallback to treating `t` as a plain `Thunk`.
        add!!(t.val, x)
```

Reply:
Are you thinking it's clearer, or allowing for better optimisation of some case?


For clarity, I think I'd vote for one less step before the action, but not a huge deal.

Reply:
I just wanted to know your thoughts on it.

I think your argument that one less step is clearer.